### PR TITLE
Feature/fga/message bubbles

### PR DESCRIPTION
**Pull request by ganfra**
This PR introduces message bubbles in Timeline.
It allows to switch Layout at runtime from the Preferences.

It also improves a bit Code Block rendering and Url previews.

### CODE REVIEW

**Comment by AnisTaluqdar** [comment]
When I send a poll my name is on the left side. It's like someone sends me a poll.

**Comment by bmarty** [comment]
(Could be nice to add some screenshots in the PR)

File: vector/src/main/res/xml/vector_settings_preferences.xml [hunk 1]
```diff
@@ -22,6 +22,11 @@
             android:title="@string/settings_theme"
             app:iconSpaceReserved="false" />
 
+        <im.vector.app.core.preference.VectorSwitchPreference
+            android:defaultValue="false"
+            android:key="SETTINGS_INTERFACE_BUBBLE_KEY"
+            android:title="@string/message_bubbles" />
```
Comment (bmarty) [inline vector/src/main/res/xml/vector_settings_preferences.xml:28]: Better to move this in the `timeline` section

**Comment by bmarty** [comment]
❌ Error: 'layout_(.*)="@\+id' detected 1 time, in file './vector/src/main/res/layout/view_message_bubble.xml', at line 203.
Rule: Use `@id` and not `@+id` when referencing ids in layouts

Also have you handled all my remarks above?

File: vector/src/main/java/im/vector/app/features/home/room/detail/timeline/url/PreviewUrlView.kt [hunk 2]
```diff
@@ -76,6 +81,22 @@ class PreviewUrlView @JvmOverloads constructor(
         }
     }
 
+    override fun render(messageLayout: TimelineMessageLayout) {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/home/room/detail/timeline/url/PreviewUrlView.kt:84]: weird to have to `render` fun in this class now

File: vector/src/main/res/layout/view_thread_room_summary.xml [hunk 3]
```diff
@@ -55,5 +55,5 @@
         app:layout_constraintHorizontal_bias="0"
         app:layout_constraintStart_toEndOf="@id/messageThreadSummaryAvatarImageView"
         app:layout_constraintTop_toTopOf="parent"
-        tools:text="@sample/messages.json/data/message" />
+        tools:text="Pouet" />
```
Comment (bmarty) [inline vector/src/main/res/layout/view_thread_room_summary.xml:58]: :) (maybe revert)

**Review by bmarty — COMMENTED** [review]
Nice work.
Maybe the changes regarding code block rendering, and URL preview management (with width and heigth) could have been handled in dedicated PR.
Still some small remarks and existing one not yet handle.
After we could merge on develop if the PR is ready!

File: vector/src/main/java/im/vector/app/features/home/room/detail/timeline/item/BaseEventItem.kt [hunk 4]
```diff
@@ -43,6 +44,20 @@ abstract class BaseEventItem<H : BaseEventItem.BaseHolder> : VectorEpoxyModel<H>
...
+    final override fun getViewType(): Int {
+        // This makes sure we have a unique integer for the combination of layout and ViewStubId.
+        return pairingFunction(layout, getViewStubId()).also {
+            Timber.v("GetViewType: for ${javaClass.canonicalName} $it with layout:$layout and stubId:${getViewStubId()}")
+        }
+    }
+
+    abstract fun getViewStubId(): Int
+
+    // Szudzik function
```
[the remark this answers is not in the document]
Comment (ganfra) [reply vector/src/main/java/im/vector/app/features/home/room/detail/timeline/item/BaseEventItem.kt:50]: Yeah I changed to Long and then normalize by substracting Int.MAX_VALUE

File: vector/src/main/res/xml/vector_settings_preferences.xml (hunk 1 above)
Comment (ganfra) [reply vector/src/main/res/xml/vector_settings_preferences.xml]: done

File: vector/src/main/java/im/vector/app/features/settings/VectorPreferences.kt [hunk 5]
```diff
@@ -852,6 +853,15 @@ class VectorPreferences @Inject constructor(private val context: Context) {
         return defaultPrefs.getBoolean(SETTINGS_SHOW_EMOJI_KEYBOARD, true)
     }
 
+    /**
+     * Tells if the emoji keyboard button should be visible or not.
+     *
+     * @return true to show emoji keyboard button.
+     */
+    fun useMessageBubblesLayout(): Boolean {
```
[the remark this answers is not in the document]
Comment (ganfra) [reply vector/src/main/java/im/vector/app/features/settings/VectorPreferences.kt:869]: Done

File: vector/src/main/java/im/vector/app/features/home/room/detail/timeline/url/PreviewUrlView.kt (hunk 2 above)
Comment (ganfra) [reply vector/src/main/java/im/vector/app/features/home/room/detail/timeline/url/PreviewUrlView.kt:84]: Changed to renderMessageLayout

**Review by bmarty — COMMENTED** [review]
Thanks for the update. Can you add a file for the changelog please?

File: vector/src/main/java/im/vector/app/features/home/room/detail/timeline/style/TimelineMessageLayoutFactory.kt [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+    }
+
+    private val cornerRadius: Float by lazy {
+        resources.getDimensionPixelSize(R.dimen.chat_bubble_corner_radius).toFloat()
+    }
+
+    private val isRTL: Boolean by lazy {
+        val currentLocale = localeProvider.current()
+        TextUtils.getLayoutDirectionFromLocale(currentLocale) == View.LAYOUT_DIRECTION_RTL
+    }
```
[the remark this answers is not in the document]
Comment (bmarty) [reply vector/src/main/java/im/vector/app/features/home/room/detail/timeline/style/TimelineMessageLayoutFactory.kt:76]: `R.bool.is_rtl` is not working properly if the user change the language inside the app.
This have to be rework. I create an issue from here :)

File: vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/MessageItemFactory.kt [hunk 7]
```diff
@@ -225,7 +223,7 @@ class MessageItemFactory @Inject constructor(
                 .locationUrl(locationUrl)
                 .mapWidth(width)
                 .mapHeight(height)
-                .userId(informationData.senderId)
+                .userId(userId)
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/MessageItemFactory.kt:226]: CC @onurays

File: vector/src/main/java/im/vector/app/features/home/room/detail/timeline/style/TimelineMessageLayoutFactory.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (bmarty) [reply vector/src/main/java/im/vector/app/features/home/room/detail/timeline/style/TimelineMessageLayoutFactory.kt:74]: Ah, I'll close the related issue then.

**Review by bmarty — APPROVED** [review]
LGTM, thanks for the update!

**Pull request merged**